streams/src/main/java/org/apache/kafka/streams/kstream/internals/WrappingNullableUtils.java

```diff
@@ -27,20 +31,37 @@
 public class WrappingNullableUtils {
 
     @SuppressWarnings("unchecked")
-    private static <T> Deserializer<T> prepareDeserializer(final Deserializer<T> specificDeserializer, final Deserializer<?> contextKeyDeserializer, final Deserializer<?> contextValueDeserializer, final boolean isKey) {
-        Deserializer<T> deserializerToUse = specificDeserializer;
-        if (deserializerToUse == null) {
+    private static <T> Deserializer<T> prepareDeserializer(final Deserializer<T> specificDeserializer, final ProcessorContext context, final boolean isKey, final String name) {
+        final Deserializer<?> contextKeyDeserializer = context.keySerde() == null ? null : context.keySerde().deserializer();
+        final Deserializer<?> contextValueDeserializer = context.valueSerde() == null ? null : context.valueSerde().deserializer();
+        final Deserializer<T> deserializerToUse;
+
+        if (specificDeserializer == null) {
             deserializerToUse = (Deserializer<T>) (isKey ? contextKeyDeserializer : contextValueDeserializer);
+        } else {
+            deserializerToUse = specificDeserializer;
+        }
+        if (deserializerToUse == null) {
+            final String serde = isKey ? "key" : "value";
+            throw new ConfigException("Failed to create deserializers. Please specify a " + serde + " serde through produced or materialized, or set one through StreamsConfig#DEFAULT_" + serde.toUpperCase(Locale.ROOT) + "_SERDE_CLASS_CONFIG for node " + name);
```

Comment:
Following the comments from above. If we throw in `context.keySerde()` we can avoid this redundant code. Of course, we should only call `context.keySerde()` if `specificDeserializer == null` for this case.

Similar below.